### PR TITLE
HAL_ChibiOS: Port for STM32L4+ processor

### PR DESCRIPTION
**Pull request by tsubashchandar**
Added support for stm32l4+ processor

- Added scripts for hwdef generation
- Tested in actual custom hardware using stm32l4r5
- Tested all peripherals and auto pilot modes.

### CODE REVIEW

**Comment by tsubashchandar** [comment]
Tested in custom hardware prototype.

**Comment by tsubashchandar** [comment]
IMG_20230409_12244950
IMG_20230409_12231314
IMG_20230409_12230787
IMG_20230409_12224841
IMG_20230409_12222209
IMG_20230409_12221902
IMG_20230409_12220862
IMG_20230409_12220650
Screenshot_2023-04-09-12-21-34-85_92460851df6f172a4592fca41cc2d2e6

**Comment by tridge** [comment]
@tsubashchandar very nice work, thanks!
can you please include a hwdef for at least one board that uses this MCU? Then we should add that board to Tools/scripts/build_ci.sh called from .github/workflows/test_chibios.yml so that we build test it in CI. Otherwise it could bitrot.

**Comment by tridge** [comment]
also, please refactor the commits with no merges and with "HAL_ChibiOS: " prefix on first line of commits. We use "Subsystem: " convention in all commits

File: libraries/AP_HAL_ChibiOS/hwdef/common/flash.c [hunk 1]
```diff
@@ -753,7 +766,7 @@ static bool stm32_flash_write_f1(uint32_t addr, const void *buf, uint32_t count)
 }
 #endif // STM32F1 or STM32F3
 
-#if defined(STM32G4) || defined(STM32L4)
+#if defined(STM32G4) || defined(STM32L4) ||  defined(STM32L4PLUS)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/flash.c:769]: extra whitespace following || - also in other areas the same thing

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py [hunk 2]
```diff
@@ -767,6 +767,15 @@ def get_flash_pages_sizes():
           mcu_series.startswith('STM32G4') or
           mcu_series.startswith('STM32L4')):
         return [ 2 ] * (get_config('FLASH_SIZE_KB', type=int)//2)
+    elif mcu_series.startswith('STM32L4'):
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:770]: I don't see how this ever gets hit give the STM32L4 above

**Review by andyp1per — CHANGES_REQUESTED** [review]
Just some formatting issues and one possible bug

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py (hunk 2 above)
Comment (tsubashchandar) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:770]: Yes Noted. Thanks !!! Mistake during upload.

File: libraries/AP_HAL_ChibiOS/hwdef/common/flash.c [hunk 3]
```diff
@@ -464,6 +467,16 @@ bool stm32_flash_erasepage(uint32_t page)
     // there is an 8th bit
     FLASH->CR |= page<<FLASH_CR_PNB_Pos;
     FLASH->CR |= FLASH_CR_STRT;
+#elif defined(STM32L4PLUS)
+    FLASH->CR |= FLASH_CR_PER;
+    if(page >= 256)
```
[the remark this answers is not in the document]
Comment (tsubashchandar) [reply libraries/AP_HAL_ChibiOS/hwdef/common/flash.c:472]: Ok

**Comment by tsubashchandar** [comment]
@tridge @andyp1per , I have fixed the review comments. But did i mess up something while rebasing and upload.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@tsubashchandar this looks good to me now (I pushed a couple of fixes to the branch), but I'd like @andyp1per to take a look too

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py [hunk 4]
```diff
@@ -762,21 +762,14 @@ def get_flash_pages_sizes():
         return [ 128 ] * (get_config('FLASH_SIZE_KB', type=int)//128)
     elif mcu_series.startswith('STM32F100') or mcu_series.startswith('STM32F103'):
         return [ 1 ] * get_config('FLASH_SIZE_KB', type=int)
+    elif mcu_series.startswith('STM32L4') and mcu_type.startswith('STM32L4R'):
+        # STM32L4PLUS
```
Comment (tsubashchandar) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:766]: @tridge It is STM32L4P.

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:766]: the mcu_type comes from this line: ```
MCU STM32L4xx STM32L4R5xx
```
is that a typo? Did you mean it to be L4P? If so then the py file needs changing too

Comment (tsubashchandar) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:766]: Yeah ok @tridge i forgot

**Review by peterbarker — DISMISSED** [review]
@tsubashchandar we don't permit merge commits into our master branch.

Did you want me to fix that for you?  It's a rebase on master that needs to be done.

**Comment by tridge** [comment]
> we don't permit merge commits into our master branch.

I've rebased it using the web UI

**Pull request merged**